designs/2019-worker-threads/README.md

```diff
@@ -0,0 +1,293 @@
+    at C:\Users\t-nagashima.AD\dev\eslint\lib\linter\safe-emitter.js:45:58
+    at Array.forEach (<anonymous>)
+    at Object.emit (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\safe-emitter.js:45:38)
+    at NodeEventGenerator.applySelector (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:253:26)
+    at NodeEventGenerator.applySelectors (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:282:22)
+    at NodeEventGenerator.enterNode (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:296:14)
+    at CodePathAnalyzer.enterNode (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\code-path-analysis\code-path-analyzer.js:646:23)
+    at C:\Users\t-nagashima.AD\dev\eslint\lib\linter\linter.js:936:32
+    at Array.forEach (<anonymous>)
+```
+
+As a side note, master terminates the other workers if a worker reported an error.
+
+#### Locally Concurrency
+
+Now ESLint does linting in parallel by thread workers if target files are many.
+
+Apart from that, each worker does linting concurrently by non-blocking IO to use CPU resource efficiently. This is a simple read-lint-report loop with `Promise.all()`:
+
+```js
+// Wait for all loops finish.
+await Promise.all(
+  // Run the loop concurrently.
+  initialFiles.map(async initialFilePath => {
+    let filePath = initialFilePath
+
+    do {
+      // Read file content by non-blocking IO.
+      const text = await readFile(filePath, "utf8")
+      // Lint the file.
+      const result = verifyText(engine, filePath, text)
+      // Send the lint result and get the next file path. This is also non-blocking async op.
+      filePath = await pushResultAndGetNextFile(result)
+    } while (filePath)
+  }),
+)
+```
+
+This logic is used in two places:
+
+1. Each worker use it.
+1. Master uses it in the main thread if target files are not many.
+
+Therefore, if `concurrency` is `1` then:
+
+```
+Main thread
+├ 🔁 Read-lint-report loop
+├ 🔁 Read-lint-report loop
+├ ...
+└ 🔁 Read-lint-report loop
+```
+
+And if `concurrency` is greater than `1` then:
+
+```
+Main thread
+├ ⏩ Worker thread
+│    ├ 🔁 Read-lint-report loop
+│    ├ ...
+│    └ 🔁 Read-lint-report loop
+├ ⏩ Worker thread
+│    ├ 🔁 Read-lint-report loop
+│    ├ ...
+│    └ 🔁 Read-lint-report loop
+├ ...
+└ ⏩ Worker thread
+      ├ 🔁 Read-lint-report loop
+      ├ ...
+      └ 🔁 Read-lint-report loop
+```
+
+#### Loading Config
+
+On the other hand, this RFC doesn't change the logic that loads config files. It's still synchronous. Because the logic is large and the configs are cached in most cases.
+
+We can improve config loading by async in the future, but it's out of the scope of this RFC.
+
+#### Constants
+
+This RFC contains two constants.
+
+- `128` ... the denominator to estimate the proper `concurrency` option value.
+- `16` ... concurrency for non-blocking IO.
+
+Those values are changeable by measuring performance in real.
```

Comment:
`measuring performance in real`? This feels to me like this should be part of a reasoning behind why these values were picked, which could include thoughts on how to benchmark which values would be the proper values.